portal-impl/src/com/liferay/portal/events/ServicePreAction.java

```diff
@@ -908,11 +908,11 @@ private ThemeDisplay _initThemeDisplay(
 		}
 
 		if (!user.isActive()) {
-			user = company.getGuestUser();
-
-			if (realUser == null) {
+			if (realUserId == user.getUserId()) {
 				httpSession.invalidate();
-			}
+			} else {
+				user = company.getGuestUser();
```

Comment:
...However, there is another problem that the "_If you are not Firstname Lastname, log out and try again._” message is incorrect because with setting the `user` the Guest user the message will show the Guest user data. I guess, this is why you've put this to the `else` branch, right?
I believe with this change the message gets also correct in case of when _there is no_ impersonation, right?

However, even if so, this message appears also in case of impersonation when I think it's not reasonable. We need to solve the case when an impersonated user has been inactivated with this message problem together, right?

If so, what if we reset the impersonation? With making the user who impersonates be the `user`. With that the message would not appear, jus the currently visited page would be shown as the impersonator would see without an incorrect message.

In the same time we could show a warning toast message with that the impersonation has been broken because of that the impersonated user has been deactivated in the meantime. The toast message could be stored here also.

In this way I believe we would not need to change the other `PortalRequestProcessor.java` either.

@stian-sigvartsen , do you think it would be a good approach?

Reply:
> ... what if we reset the impersonation? With making the user who impersonates be the user. With that the message would not appear, jus the currently visited page would be shown as the impersonator would see without an incorrect message.

In the first times I tried to reset the impersonation to prevent this error related to the guest user, but it didn't work as the user is still "stuck" with the request parameter `doAsUserId` when I refresh the page and the user menu disappears as well. 

I had added this block. I also needed to redefine it in the request because the user was stuck in the terms and conditions page without it.
```
if (!user.isActive()) {
			if (realUserId == user.getUserId()) {
				httpSession.invalidate();
			} else {
				user = realUser;
				httpServletRequest.setAttribute(WebKeys.USER, realUser);
			} 
		}
```

Any ideas how we could bypass or remove this request attribute?

Reply:
Modifying `ServicePreAction`. or `PortalRequestProcessor` to handle the impersonation scenario can be risky and make things more fragile.

I think you should consider a lighter touch for this solution. We are already handling gracefully situations when someone provides an invalid `doAsUserId` parameter value on the querystring, so we can piggy back on that. I would add a check for the user's active status after this line: https://github.com/liferay/liferay-portal/blob/3414ee92ce4e20e29f81584ae8b82b0fdd63f154/portal-impl/src/com/liferay/portal/util/PortalImpl.java#L7118 

This could be combined with removing the "Impersonate" option from the actions menu of an inactive user. Then a admin will not get into the situation of attempting to impersonate an inactive user in the first place so we don't need further messaging.

What do you think?

Reply:
Hi @stian-sigvartsen 

Thank you very much for your insight, It was very helpful. I added the verification where you suggested and now it's working as expected. Please check: https://github.com/liferay-appsec/liferay-portal/pull/1412/commits/85bcda6542cc58fdbf6cb47110220e159872ea74

Reply:
Thank you !